types/meteor/mongo.d.ts

```diff
@@ -128,44 +128,46 @@ declare module "meteor/mongo" {
             [id: string]: Number;
         }
 
+        type TransformFn<T> = ((doc: T) => any) | null;
+
+        type Options<T> = {
+            sort?: SortSpecifier;
+            skip?: number;
+            limit?: number;
+            fields?: FieldSpecifier;
+            reactive?: boolean;
+            transform?: TransformFn<T>;
+        }
+
+        type DispatchTransform<Transf, DefType, U> = Transf extends (...args: any) => any ? ReturnType<Transf> : Transf extends null ? DefType : U;
+
         var Collection: CollectionStatic;
         interface CollectionStatic {
-            new <T>(name: string | null, options?: {
+            new <T, U = T>(name: string | null, options?: {
                 connection?: Object | null;
                 idGeneration?: string;
-                transform?: Function | null;
-            }): Collection<T>;
-        }
-        interface Collection<T> {
-            allow(options: {
-                insert?: (userId: string, doc: T) => boolean;
-                update?: (userId: string, doc: T, fieldNames: string[], modifier: any) => boolean;
-                remove?: (userId: string, doc: T) => boolean;
+                transform?: (doc: T) => U;
+            }): Collection<T, U>;
+        }
+        interface Collection<T, U = T> {
+            allow<Fn extends TransformFn<T> | undefined = undefined, Doc = DispatchTransform<Fn, T, U>>(options: {
+                insert?: (userId: string, doc: Doc) => boolean;
+                update?: (userId: string, doc: Doc, fieldNames: string[], modifier: any) => boolean;
+                remove?: (userId: string, doc: Doc) => boolean;
```

Comment:
I believe you can achieve the same simply defining `doc: U`, e.g. `insert?: (userId: string, doc: U) => boolean;`, generic on the function level not needed. or am I missing something?

Reply:
We should check transform function to null